app/lib/three_scale/spam_protection/integration/controller.rb

```diff
@@ -1,27 +1,74 @@
+# frozen_string_literal: true
+
 module ThreeScale::SpamProtection
+
   module Integration
 
     module Controller
+      extend ActiveSupport::Concern
+
+      class_methods do
+
+        def enabled_checks
+          _enabled_checks
+        end
+
+        def has_spam_protection(*checks)
+          self._enabled_checks = checks.presence || AVAILABLE_CHECKS
+        end
+      end
+
+      included do
+        delegate :spam?, :spam_probability, to: :spam_protection
+        class_attribute :_enabled_checks, instance_reader: false, instance_writer: false
+        before_action :instantiate_checks
+      end
+
+      def instantiate_checks
+        spam_protection_conf.enable_checks! self.class.enabled_checks
+      end
+
+      def spam_protection
+        @spam_protection ||= ThreeScale::SpamProtection::Protector.new(self)
+      end
+
+      def spam_protection_conf
+        @spam_protection_conf ||= ThreeScale::SpamProtection::Configuration.new(self)
+      end
+
+      def spam_protection_form(form)
+        ThreeScale::SpamProtection::FormProtector.new(form, self)
+      end
+
+      def store
+        @store ||= SessionStore.new(request.session)
+      end
+
+      def level
+        site_account.settings.spam_protection_level
+      end
+
       private
 
-      def spam_check_save(object, &block)
-        block_given? ? yield : object.save if spam_check(object)
+      def spam_check_save(object)
+        (block_given? ? yield : object.save) if spam_check(object)
       end
 
       def verify_captcha(object)
         verify_recaptcha(model: object, attribute: :recaptcha)
       end
 
-      def spam_check(object, session_store: ThreeScale::SpamProtection::SessionStore)
-        return true if logged_in?
-
-        case level = site_account.settings.spam_protection_level
-
+      # Called when a form is received. Usually through a PUT or POST request.
+      # Checks the received data to try to detect bots.
+      #
+      # FormProtector#to_str will read the result of this call in the next GET request to load the form.
+      def spam_check(object)
+        case level
         when :none
           true
         when :auto
-          if object.spam?
-            session_store.new(request.session).mark_possible_spam
+          if spam?
+            @store.mark_possible_spam
             Rails.logger.debug "[SpamProtection][Integration] Captcha filled and object is spam - verifying captcha"
```

Comment:
I know it has been already like that, but I think this log is not always true... 
It is printed even when there was no captcha shown to the user, but the spam checks detected a risk. But well, it's OK.

Reply:
Yes, it's a bit confusing. In `:auto` mode two things can happen: receiving the check inputs, or not receiving them because a captcha was shown in their place. In both cases `spam?` returns `true`. So the logic is correct, but the message is confusing. I reworded it to `Spam detected or captcha filled - verifying captcha`. I hope that's better.